p2p/proto/snapshot.proto

```diff
@@ -0,0 +1,82 @@
+syntax = "proto3";
+
+import "p2p/proto/common.proto";
+import "p2p/proto/state.proto";
+
+message PatriciaNode {
+    message Edge {
+        uint32 length = 1;
+        Felt252 path = 2; // as bits of left/right
+        Felt252 value = 3;
+    }
+    message Binary {
+        Felt252 left = 1;
+        Felt252 right = 2;
+    }
+
+    oneof node {
+        Edge edge = 1;
+        Binary binary = 2;
+    }
+}
+
+// non leaf nodes required to build the trie given the range (leaves)
+message PatriciaRangeProof {
+    repeated PatriciaNode nodes = 1;
+}
+
+// leafs of the contract state tre
+message ContractState {
+    Hash address = 1; // the key
+    Hash class = 2;
+    Hash storage = 3; // patricia
+    uint64 nonce = 4;
+}
+
+// request a range from the contract state tree that matches the given root (block)
+// starts at 'start' and ends no less than 'end'.
+// the result is  (ContractRange+, PatriciaRangeProof)*
+message GetContractRange {
+    Hash root = 2;
+    Hash start = 3;
+    Hash end = 4;
+    uint32 chunks_per_proof = 5; // how many ContractRange items to send before sending a proof
+}
+
+// stream of leaves in the contracts tree
+message ContractRange {
+    repeated ContractState state = 1;
+}
+
+// duplicate of GetContractRange. Can introduce a 'type' instead.
+// result is (Classes+, PatriciaRangeProof)*
+message GetClassRange {
+    Hash root = 1;
+    Hash start = 2;
+    Hash end = 3;
+    uint32 chunks_per_proof = 4;
+}
+
+// A position in some contract's state tree is identified by the state tree's root and the key in it
+message StorageLeafQuery {
+    Hash root = 1;
+    Felt252 key = 2;
+}
+
+message StorageRangeQuery {
+    uint32 range_id = 1;
```

Comment:
What is a `range_id`?

Reply:
Something that is echoed back in the reply. So the client can track which response matches each query.

Reply:
Yea..  I recommend just creating a new libp2p stream for each request-response pair so this is not needed, makes implementation a whole lot easier. Just formalize that libp2p must be setup with yamux multiplexer so that it uses the same tcp connection internally.

Reply:
Yea..  I recommend just creating a new libp2p stream for each request-response pair so this is not needed, makes implementation a whole lot easier. Just formalize that libp2p must be setup with yamux multiplexer so that it uses the same tcp connection internally.

Reply:
Ah I think https://github.com/starknet-io/starknet-p2p-specs/pull/8/files#r1293631721 is there for the same reason?

> I recommend just creating a new libp2p stream for each request-response pair so this is not needed, makes implementation a whole lot easier

Agreed

Reply:
> I recommend just creating a new libp2p stream for each request-response pair so this is not needed, makes implementation a whole lot easier

I will be going this route

Reply:
Will look into this

Reply:
So if we're doing a 1:1 between streams and requests, then there is no need for the Fin message. The stream just closes reading the last message in the response.